sentence_transformers/SentenceTransformer.py

```diff
@@ -175,27 +181,37 @@ def __init__(
                     # A model from sentence-transformers
                     model_name_or_path = __MODEL_HUB_ORGANIZATION__ + "/" + model_name_or_path
 
-            if is_sentence_transformer_model(model_name_or_path, token, cache_folder=cache_folder, revision=revision):
-                modules = self._load_sbert_model(
+            hub_kwargs = {
+                "revision": revision,
+                "token": token,
+                "trust_remote_code": trust_remote_code,
+                "cache_dir": cache_folder,
+            }
+
+            if model_or_path_has_file("modules.json", model_name_or_path, **hub_kwargs):
+                # TODO: kwargs are ignored, warn?
+                modules = self._load_sbert_model_as_modules(model_name_or_path, **hub_kwargs)
+            elif model_or_path_has_file("config_sentence_transformers.json", model_name_or_path, **hub_kwargs):
+                self.config: SentenceTransformerConfig = SentenceTransformerConfig.from_pretrained(
                     model_name_or_path,
-                    token=token,
-                    cache_folder=cache_folder,
-                    revision=revision,
-                    trust_remote_code=trust_remote_code,
+                    _configuration_file="config_sentence_transformers.json",
+                    **hub_kwargs,
+                    loaded_with_from_pretrained=True,
                 )
+                modules = self._load_sbert_model_as_pretrained_model(model_name_or_path, **hub_kwargs, **kwargs)
             else:
-                modules = self._load_auto_model(
-                    model_name_or_path,
-                    token=token,
-                    cache_folder=cache_folder,
-                    revision=revision,
-                    trust_remote_code=trust_remote_code,
-                )
+                # TODO: pass kwargs as model_args to Transformer?
+                modules = self._load_auto_model(model_name_or_path, **hub_kwargs)
 
         if modules is not None and not isinstance(modules, OrderedDict):
             modules = OrderedDict([(str(idx), module) for idx, module in enumerate(modules)])
 
+        if self.config is None:
+            self.config = SentenceTransformerConfig()
+
+        self.calling_init = True
```

Comment:
It's a bit hard to follow around why we need this `calling_init` and if it wouldn't be simpler to pass arguments around rather than keep a state of the init calls